// webpay/resources/charge.js
var dataTypes = require('../data_types');

function Charge(client) {
  this.client = client;
}

Charge.prototype.create = function create(params, callback) {
  if (typeof params === 'function') {
    callback = params;
    params = {};
  }
  var req = dataTypes.ChargeRequestCreate.create(params);
  if (req instanceof Error) {
    return callback(req, null);
  }
  this.client.request('post', 'charges', req, function(err, data) {
    if (err) {
      callback(err, null);
    } else {
      var res = new dataTypes.ChargeResponse(data);
      if (res instanceof Error) {
        callback(res, null);
      } else {
        callback(null, res);
      }
    }
  });
};

Charge.prototype.retrieve = function retrieve(params, callback) {
  if (typeof params === 'function') {
    callback = params;
    params = {};
  }
  var req = dataTypes.ChargeIdRequest.create(params);
  if (req instanceof Error) {
    return callback(req, null);
  }
  this.client.request('get', 'charges' + '/' + req.id.toString(), req, function(err, data) {
    if (err) {
      callback(err, null);
    } else {
      var res = new dataTypes.ChargeResponse(data);
      if (res instanceof Error) {
        callback(res, null);
      } else {
        callback(null, res);
      }
    }
  });
};

Charge.prototype.refund = function refund(params, callback) {
  if (typeof params === 'function') {
    callback = params;
    params = {};
  }
  var req = dataTypes.ChargeRequestWithAmount.create(params);
  if (req instanceof Error) {
    return callback(req, null);
  }
  this.client.request('post', 'charges' + '/' + req.id.toString() + '/' + 'refund', req, function(err, data) {
    if (err) {
      callback(err, null);
    } else {
      var res = new dataTypes.ChargeResponse(data);
      if (res instanceof Error) {
        callback(res, null);
      } else {
        callback(null, res);
      }
    }
  });
};

Charge.prototype.capture = function capture(params, callback) {
  if (typeof params === 'function') {
    callback = params;
    params = {};
  }
  var req = dataTypes.ChargeRequestWithAmount.create(params);
  if (req instanceof Error) {
    return callback(req, null);
  }
  this.client.request('post', 'charges' + '/' + req.id.toString() + '/' + 'capture', req, function(err, data) {
    if (err) {
      callback(err, null);
    } else {
      var res = new dataTypes.ChargeResponse(data);
      if (res instanceof Error) {
        callback(res, null);
      } else {
        callback(null, res);
      }
    }
  });
};

Charge.prototype.all = function all(params, callback) {
  if (typeof params === 'function') {
    callback = params;
    params = {};
  }
  var req = dataTypes.ChargeListRequest.create(params);
  if (req instanceof Error) {
    return callback(req, null);
  }
  this.client.request('get', 'charges', req, function(err, data) {
    if (err) {
      callback(err, null);
    } else {
      var res = new dataTypes.ChargeResponseList(data);
      if (res instanceof Error) {
        callback(res, null);
      } else {
        callback(null, res);
      }
    }
  });
};

module.exports = Charge;
