// webpay/resources/customer.js
var dataTypes = require('../data_types');

function Customer(client) {
  this.client = client;
}

Customer.prototype.create = function create(params, callback) {
  if (typeof params === 'function') {
    callback = params;
    params = {};
  }
  var req = dataTypes.CustomerRequestCreate.create(params);
  if (req instanceof Error) {
    return callback(req, null);
  }
  this.client.request('post', 'customers', req, function(err, data) {
    if (err) {
      callback(err, null);
    } else {
      var res = new dataTypes.CustomerResponse(data);
      if (res instanceof Error) {
        callback(res, null);
      } else {
        callback(null, res);
      }
    }
  });
};

Customer.prototype.retrieve = function retrieve(params, callback) {
  if (typeof params === 'function') {
    callback = params;
    params = {};
  }
  var req = dataTypes.CustomerIdRequest.create(params);
  if (req instanceof Error) {
    return callback(req, null);
  }
  this.client.request('get', 'customers' + '/' + req.id.toString(), req, function(err, data) {
    if (err) {
      callback(err, null);
    } else {
      var res = new dataTypes.CustomerResponse(data);
      if (res instanceof Error) {
        callback(res, null);
      } else {
        callback(null, res);
      }
    }
  });
};

Customer.prototype.update = function update(params, callback) {
  if (typeof params === 'function') {
    callback = params;
    params = {};
  }
  var req = dataTypes.CustomerRequestUpdate.create(params);
  if (req instanceof Error) {
    return callback(req, null);
  }
  this.client.request('post', 'customers' + '/' + req.id.toString(), req, function(err, data) {
    if (err) {
      callback(err, null);
    } else {
      var res = new dataTypes.CustomerResponse(data);
      if (res instanceof Error) {
        callback(res, null);
      } else {
        callback(null, res);
      }
    }
  });
};

Customer.prototype.delete = function delete_(params, callback) {
  if (typeof params === 'function') {
    callback = params;
    params = {};
  }
  var req = dataTypes.CustomerIdRequest.create(params);
  if (req instanceof Error) {
    return callback(req, null);
  }
  this.client.request('delete', 'customers' + '/' + req.id.toString(), req, function(err, data) {
    if (err) {
      callback(err, null);
    } else {
      var res = new dataTypes.CustomerResponse(data);
      if (res instanceof Error) {
        callback(res, null);
      } else {
        callback(null, res);
      }
    }
  });
};

Customer.prototype.all = function all(params, callback) {
  if (typeof params === 'function') {
    callback = params;
    params = {};
  }
  var req = dataTypes.BasicListRequest.create(params);
  if (req instanceof Error) {
    return callback(req, null);
  }
  this.client.request('get', 'customers', req, function(err, data) {
    if (err) {
      callback(err, null);
    } else {
      var res = new dataTypes.CustomerResponseList(data);
      if (res instanceof Error) {
        callback(res, null);
      } else {
        callback(null, res);
      }
    }
  });
};

Customer.prototype.deleteActiveCard = function deleteActiveCard(params, callback) {
  if (typeof params === 'function') {
    callback = params;
    params = {};
  }
  var req = dataTypes.CustomerIdRequest.create(params);
  if (req instanceof Error) {
    return callback(req, null);
  }
  this.client.request('delete', 'customers' + '/' + req.id.toString() + '/' + 'active_card', req, function(err, data) {
    if (err) {
      callback(err, null);
    } else {
      var res = new dataTypes.CustomerResponse(data);
      if (res instanceof Error) {
        callback(res, null);
      } else {
        callback(null, res);
      }
    }
  });
};

module.exports = Customer;
