// webpay/resources/recursion.js
var dataTypes = require('../data_types');

function Recursion(client) {
  this.client = client;
}

Recursion.prototype.create = function create(params, callback) {
  if (typeof params === 'function') {
    callback = params;
    params = {};
  }
  var req = dataTypes.RecursionRequestCreate.create(params);
  if (req instanceof Error) {
    return callback(req, null);
  }
  this.client.request('post', 'recursions', req, function(err, data) {
    if (err) {
      callback(err, null);
    } else {
      var res = new dataTypes.RecursionResponse(data);
      if (res instanceof Error) {
        callback(res, null);
      } else {
        callback(null, res);
      }
    }
  });
};

Recursion.prototype.retrieve = function retrieve(params, callback) {
  if (typeof params === 'function') {
    callback = params;
    params = {};
  }
  var req = dataTypes.RecursionIdRequest.create(params);
  if (req instanceof Error) {
    return callback(req, null);
  }
  this.client.request('get', 'recursions' + '/' + req.id.toString(), req, function(err, data) {
    if (err) {
      callback(err, null);
    } else {
      var res = new dataTypes.RecursionResponse(data);
      if (res instanceof Error) {
        callback(res, null);
      } else {
        callback(null, res);
      }
    }
  });
};

Recursion.prototype.resume = function resume(params, callback) {
  if (typeof params === 'function') {
    callback = params;
    params = {};
  }
  var req = dataTypes.RecursionRequestResume.create(params);
  if (req instanceof Error) {
    return callback(req, null);
  }
  this.client.request('post', 'recursions' + '/' + req.id.toString() + '/' + 'resume', req, function(err, data) {
    if (err) {
      callback(err, null);
    } else {
      var res = new dataTypes.RecursionResponse(data);
      if (res instanceof Error) {
        callback(res, null);
      } else {
        callback(null, res);
      }
    }
  });
};

Recursion.prototype.delete = function delete_(params, callback) {
  if (typeof params === 'function') {
    callback = params;
    params = {};
  }
  var req = dataTypes.RecursionIdRequest.create(params);
  if (req instanceof Error) {
    return callback(req, null);
  }
  this.client.request('delete', 'recursions' + '/' + req.id.toString(), req, function(err, data) {
    if (err) {
      callback(err, null);
    } else {
      var res = new dataTypes.RecursionResponse(data);
      if (res instanceof Error) {
        callback(res, null);
      } else {
        callback(null, res);
      }
    }
  });
};

Recursion.prototype.all = function all(params, callback) {
  if (typeof params === 'function') {
    callback = params;
    params = {};
  }
  var req = dataTypes.RecursionListRequest.create(params);
  if (req instanceof Error) {
    return callback(req, null);
  }
  this.client.request('get', 'recursions', req, function(err, data) {
    if (err) {
      callback(err, null);
    } else {
      var res = new dataTypes.RecursionResponseList(data);
      if (res instanceof Error) {
        callback(res, null);
      } else {
        callback(null, res);
      }
    }
  });
};

module.exports = Recursion;
