// webpay/resources/shop.js
var dataTypes = require('../data_types');

function Shop(client) {
  this.client = client;
}

Shop.prototype.create = function create(params, callback) {
  if (typeof params === 'function') {
    callback = params;
    params = {};
  }
  var req = dataTypes.ShopRequestCreate.create(params);
  if (req instanceof Error) {
    return callback(req, null);
  }
  this.client.request('post', 'shops', req, function(err, data) {
    if (err) {
      callback(err, null);
    } else {
      var res = new dataTypes.ShopResponse(data);
      if (res instanceof Error) {
        callback(res, null);
      } else {
        callback(null, res);
      }
    }
  });
};

Shop.prototype.retrieve = function retrieve(params, callback) {
  if (typeof params === 'function') {
    callback = params;
    params = {};
  }
  var req = dataTypes.ShopIdRequest.create(params);
  if (req instanceof Error) {
    return callback(req, null);
  }
  this.client.request('get', 'shops' + '/' + req.id.toString(), req, function(err, data) {
    if (err) {
      callback(err, null);
    } else {
      var res = new dataTypes.ShopResponse(data);
      if (res instanceof Error) {
        callback(res, null);
      } else {
        callback(null, res);
      }
    }
  });
};

Shop.prototype.update = function update(params, callback) {
  if (typeof params === 'function') {
    callback = params;
    params = {};
  }
  var req = dataTypes.ShopRequestUpdate.create(params);
  if (req instanceof Error) {
    return callback(req, null);
  }
  this.client.request('post', 'shops' + '/' + req.id.toString(), req, function(err, data) {
    if (err) {
      callback(err, null);
    } else {
      var res = new dataTypes.ShopResponse(data);
      if (res instanceof Error) {
        callback(res, null);
      } else {
        callback(null, res);
      }
    }
  });
};

Shop.prototype.all = function all(params, callback) {
  if (typeof params === 'function') {
    callback = params;
    params = {};
  }
  var req = dataTypes.BasicListRequest.create(params);
  if (req instanceof Error) {
    return callback(req, null);
  }
  this.client.request('get', 'shops', req, function(err, data) {
    if (err) {
      callback(err, null);
    } else {
      var res = new dataTypes.ShopResponseList(data);
      if (res instanceof Error) {
        callback(res, null);
      } else {
        callback(null, res);
      }
    }
  });
};

module.exports = Shop;
